File: PythonSpark/Aggregations/Views/ViewsAggregationJob.py
from abc import ABC

from pyspark.sql import DataFrame

from Aggregations.AggregationJob import AggregationJob
from Utils.Loading import load_dynamic_table, load_static_table
from Utils.Parsing import filter_data_frame
from Utils.RichArgumentParser import RichArgumentParser


class ViewsAggregationJob(AggregationJob, ABC):
    def load(self) -> dict:
        return {
            "views": load_dynamic_table("views", self.arguments, self.spark),
            "ips": load_static_table("ips", self.arguments, self.spark)
        }

    def process(self, data: dict) -> DataFrame:
        joined = data["views"].join(data["ips"], "ip")
        return filter_data_frame(joined,
                                 self.arguments.locations,
                                 self.arguments.time_frame,
                                 self.arguments.devices)

    @staticmethod
    def parser() -> RichArgumentParser:
        return AggregationJob.parser() \
            .add_dynamic_table_data_source("views") \
            .add_static_table_data_source("ips") \
            .add_locations() \
            .add_devices() \
            .add_time_frame()

File: ScalaSpark/src/main/scala/Aggregations/Views/ViewsAggregationJob.scala
package Aggregations.Views

import Aggregations.AggregationJob
import JobManagment.JobCompanion
import Utils.ArgumentsParsing.RichParser
import Utils.Loading.{loadDynamic, loadStatic}
import Utils.Parsing.filterData
import de.halcony.argparse.Parser
import org.apache.spark.sql.{DataFrame, SparkSession}

abstract class ViewsAggregationJob(arguments: Map[String, String], spark: SparkSession) extends AggregationJob(arguments, spark) {
  override def load(): Map[String, DataFrame] = {
    Map(
      "views" -> loadDynamic("views", arguments, spark),
      "ips" -> loadStatic("ips", arguments, spark)
    )
  }

  override def process(data: Map[String, DataFrame]): DataFrame = {
    val viewsWithLocations = data("views").join(data("ips"), "ip")

    filterData(arguments("timeFrame"), arguments("devices"), arguments("locations"), viewsWithLocations)
  }
}

object ViewsAggregationJob extends JobCompanion {
  override def parser(): Parser = {
    AggregationJob.parser()
      .addDynamicTableDataSource("views")
      .addStaticTableDataSource("ips")
      .addTimeFrame()
      .addDevices()
      .addLocation()
  }
}